### PR TITLE
Add route building control

### PR DESCRIPTION
**Pull request by matthew44-mappable**
Fixes #

### CODE REVIEW

File: src/controls/MMapRouteControl/index.css [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    gap: 20px;
+    border-radius: 12px;
+    border: 1px solid #ebedf0;
+    background: #fff;
+
+    box-shadow:
+        0px 4px 12px 0px rgba(95, 105, 131, 0.1),
+        0px 4px 24px 0px rgba(95, 105, 131, 0.04);
+
+    &.error {
```
Comment (xdan) [inline src/controls/MMapRouteControl/index.css:47]: also

File: src/controls/MMapRouteControl/index.ts [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    constructor(props: MMapRouteControlProps) {
+        super(props);
+    }
+
+    protected _onAttach(): void {
+        this._control = new mappable.MMapControl({transparent: true});
+        this._router = new MMapCommonRouteControl(this._props);
+
+        this._control.addChild(this._router);
+        this.addChild(this._control);
```
Comment (xdan) [inline src/controls/MMapRouteControl/index.ts:65]: direct

File: src/controls/MMapRouteControl/index.ts [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+                const {theme} = this._consumeContext(mappable.ThemeContext);
+                this._rootElement.classList.toggle('_dark', theme === 'dark');
+            },
+            {immediate: true}
+        );
+    }
+
+    protected _onDetach(): void {
+        this._detachDom?.();
+        this._detachDom = undefined;
```
Comment (xdan) [inline src/controls/MMapRouteControl/index.ts:141]: Again need remove children

File: src/controls/MMapRouteControl/utils.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
+const SEC_TO_DAYS = 24 * 3600;
+const SEC_TO_HOURS = 3600;
+const SEC_TO_MIN = 60;
+
+export function formatDistance(length: number): string {
```
Comment (xdan) [inline src/controls/MMapRouteControl/utils.ts:5]: metters: lengthMetters: or jsdoc

**Review by xdan — CHANGES_REQUESTED** [review]

File: src/controls/MMapRouteControl/MMapWaypointInput/index.ts [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+                this._props.waypoint = undefined;
+                this._resetInput();
+            } else {
+                this._search({text: this._props.waypoint.toString()}, this._props.waypoint);
+            }
+        }
+    }
+
+    protected _onDetach(): void {
+        this._detachDom?.();
```
[the remark this answers is not in the document]
Comment (matthew44-mappable) [reply src/controls/MMapRouteControl/MMapWaypointInput/index.ts:161]: I moved the creation of entities to the constructor

File: src/controls/MMapRouteControl/MMapWaypointInput/index.ts [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+        this._detachDom = undefined;
+    }
+
+    private _updateIndicatorStatus(status: 'empty' | 'focus' | 'setted'): void {
+        this._indicator.innerHTML = '';
+        this._indicator.classList.toggle('_empty', status === 'empty');
+
+        switch (status) {
+            case 'empty':
+                this._indicator.insertAdjacentHTML('afterbegin', emptyIndicatorSVG);
```
[the remark this answers is not in the document]
Comment (matthew44-mappable) [reply src/controls/MMapRouteControl/MMapWaypointInput/index.ts:171]: I didn't quite understand the question and what the problem was. If the input field is empty, an indicator is displayed next to it, indicating an empty input field. This was described in the UI design 🤷‍♂️

File: src/controls/MMapRouteControl/MMapWaypointInput/index.ts [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+
+    private _submitWaypointInput = (event?: SubmitEvent) => {
+        event?.preventDefault();
+        if (!this._suggestComponent.activeSuggest) {
+            this._inputEl.focus();
+            return;
+        }
+        const {uri, text} = this._suggestComponent.activeSuggest.dataset;
+        this._search({uri, text});
+        this.removeChild(this._suggestComponent);
```
[the remark this answers is not in the document]
Comment (matthew44-mappable) [reply src/controls/MMapRouteControl/MMapWaypointInput/index.ts:225]: Why? We added `_suggestComponent` via the "addChild" method

File: src/controls/MMapRouteControl/MMapWaypointInput/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (xdan) [reply src/controls/MMapRouteControl/MMapWaypointInput/index.ts:176]: No, I'm talking about insertAdjacentHTML
Why is it needed? You have an empty DOM element
Just innerHTML?

File: src/controls/MMapRouteControl/MMapWaypointInput/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (xdan) [reply src/controls/MMapRouteControl/MMapWaypointInput/index.ts:230]: use addDirectChild, What happens if your component is inserted into React? children in react will displace those you use inside

[the remark this answers is not in the document]
Comment (matthew44-mappable) [reply src/controls/MMapRouteControl/MMapWaypointInput/index.ts]: Also fixed this it in the search control

**Pull request merged**